src/ioctls/vcpu.rs

```diff
@@ -1188,8 +1197,8 @@ impl VcpuFd {
     ///
     /// * `reg_id` - ID of the register.
     #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
-    pub fn get_one_reg(&self, reg_id: u64) -> Result<u128> {
-        let mut reg_value = 0;
+    pub fn get_one_reg<T: ArmRegister>(&self, reg_id: u64) -> Result<T> {
+        let mut reg_value = std::mem::MaybeUninit::<T>::uninit();
```

Comment:
Mh, I think with this implementation the function needs to be marked `unsafe`: If I call, say, `get_one_reg::<u32>` for a register that is 64 bit wide, then we'll pass a point to a 32 bit wide value to the kernel, to which the kernel will happily write 64 bits, causing stack corruption. 

An alternative might be using https://github.com/torvalds/linux/blob/f1fcbaa18b28dec10281551dfe6ed3a3ed80e3d6/include/uapi/linux/kvm.h#L1364-L1374 to do a width-validation before doing the ioctl (e.g. `if std::mem::size_of::<T>() != register_width {return error}`). That shouldn't have too much of a performance impact (should be a single instruction or so) - although maybe we'll still want to also provide a `get_one_reg_unsafe` that doesnt do the check?

Reply:
I was thinking, what if we change `get_one_reg` to accept buffer to write register to?
So it would be:
```rust
    pub fn set_one_reg(&self, reg_id: u64, data: &[u8]) -> Result<usize> {
        let reg_size = 2_usize.pow(((reg_id & KVM_REG_SIZE_MASK) >> KVM_REG_SIZE_SHIFT) as u32);
        if data.len() < reg_size {
            return Err(errno::Error::new(libc::EINVAL));
        }
        let onereg = kvm_one_reg {
            id: reg_id,
            addr: data as *const _ as u64,
        };
        // SAFETY: This is safe because we allocated the struct and we know the kernel will read
        // exactly the size of the struct.
        let ret = unsafe { ioctl_with_ref(self, KVM_SET_ONE_REG(), &onereg) };
        if ret < 0 {
            return Err(errno::Error::last());
        }
        Ok(reg_size)
    }

    pub fn get_one_reg(&self, reg_id: u64, data: &mut [u8]) -> Result<usize> {
        let reg_size = 2_usize.pow(((reg_id & KVM_REG_SIZE_MASK) >> KVM_REG_SIZE_SHIFT) as u32);
        if data.len() < reg_size {
            return Err(errno::Error::new(libc::EINVAL));
        }
        let mut onereg = kvm_one_reg {
            id: reg_id,
            addr: data as *mut _ as u64,
        };
        // SAFETY: This is safe because we allocated the struct and we know the kernel will read
        // exactly the size of the struct.
        let ret = unsafe { ioctl_with_mut_ref(self, KVM_GET_ONE_REG(), &mut onereg) };
        if ret < 0 {
            return Err(errno::Error::last());
        }
        // SAFETY: Value is initialized by the kernel
        Ok(reg_size)
    }
```

Reply:
I like this approach a lot!

Reply:
Ok converted to this new approach.